### ISSUE TITLE
Re-usable links not linting

### ISSUE THREAD

**Issue by catharsisjelly** [wework/speccy#21]
I'm fairly new to using OpenAPI 3.0 so it's possible that I don't entirely understand it. but for some reason the re-usable links don't seem to validate properly although Swagger's editor seems to think it's fine.

## Detailed description

The following OpenAPI spec validates fine without link references, but I was wanting to make the `Link` a reference

```yaml
openapi: "3.0.1"
info:
  description: "Something to demo that Link components are not linting"
  version: "1.0.0"
  title: "Test API"
  contact:
    name: "Chris Lock"
    url: "https://catharsis.co.uk"
servers:
  - url: "https://api.not-a-real-api.com"
tags:
  - name: "administrators"
    description: "Administrator end points"
paths:
  /administrators/{userId}:
    parameters:
      - $ref: "#/components/parameters/userId"
    put:
      operationId: "assignAdministrator"
      tags:
        - "users"
        - "administrators"
      summary: "Assign an Administrator to a User"
      requestBody:
        $ref: "#/components/requestBodies/administratorRequest"
      responses:
        200:
          $ref: "#/components/responses/administratorResponse"
        404:
          $ref: "#/components/responses/404"
        405:
          $ref: "#/components/responses/405"
        410:
          $ref: "#/components/responses/410"

components:
  securitySchemes:
    oauth_auth:
      type: "oauth2"
      flows:
        clientCredentials:
          tokenUrl: "/oauth/token"
          scopes:
            "read:user": "read a users"
            "write:user": "write the users"
        password:
          tokenUrl: "/oauth/token"
          refreshUrl: "/oauth/token"
          scopes:
            "read:user": "read a users"
            "write:user": "write the users"

  parameters:
    userId:
      description: "The id of the User"
      in: "path"
      name: "userId"
      required: true
      schema:
        type: "integer"

  requestBodies:
    administratorRequest:
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/Administrator"

  responses:
    204:
      description: "No content"
    404:
      description: "Resource not found"
    405:
      description: "Invalid input"
    410:
      description: "Resource has been deleted"
    administratorResponse:
      description: "The Administrator object response"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/Administrator"
      links:
        getUserById:
          operationId: "getUser"
          parameters:
            userId: "$request.path.userId"

  schemas:
    Administrator:
      type: "object"
      properties:
        id:
          type: "integer"
          format: "int64"
          readOnly: true
        name:
          type: "string"
        mail:
          type: "string"
          format: "email"
        enabled:
          type: "boolean"

security:
  - oauth_auth: []
```

But the below does not validate using Speccy but again using the Swagger editor it seems fine

```yaml
openapi: "3.0.1"
info:
  description: "Something to demo that Link components are not linting"
  version: "1.0.0"
  title: "Test API"
  contact:
    name: "Chris Lock"
    url: "https://catharsis.co.uk"
servers:
  - url: "https://api.not-a-real-api.com"
tags:
  - name: "administrators"
    description: "Administrator end points"
paths:
  /administrators/{userId}:
    put:
      operationId: "assignAdministrator"
      tags:
        - "users"
        - "administrators"
      summary: "Assign an Administrator to a User"
      requestBody:
        $ref: "#/components/requestBodies/administratorRequest"
      responses:
        200:
          $ref: "#/components/responses/administratorResponse"
        404:
          $ref: "#/components/responses/404"
        405:
          $ref: "#/components/responses/405"
        410:
          $ref: "#/components/responses/410"

components:
  securitySchemes:
    oauth_auth:
      type: "oauth2"
      flows:
        clientCredentials:
          tokenUrl: "/oauth/token"
          scopes:
            "read:user": "read a users"
            "write:user": "write the users"
        password:
          tokenUrl: "/oauth/token"
          refreshUrl: "/oauth/token"
          scopes:
            "read:user": "read a users"
            "write:user": "write the users"

  links:
    getUserById:
      operationId: "getUser"
      parameters:
        userId: "$request.path.userId"

  parameters:
    userId:
      description: "The id of the User"
      in: "path"
      name: "userId"
      required: true
      schema:
        type: "integer"

  requestBodies:
    administratorRequest:
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/Administrator"

  responses:
    204:
      description: "No content"
    404:
      description: "Resource not found"
    405:
      description: "Invalid input"
    410:
      description: "Resource has been deleted"
    administratorResponse:
      description: "The Administrator object response"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/Administrator"
      links:
        getUserById:
          $ref: "#/components/links/getUserById"

  schemas:
    Administrator:
      type: "object"
      properties:
        id:
          type: "integer"
          format: "int64"
          readOnly: true
        name:
          type: "string"
        mail:
          type: "string"
          format: "email"
        enabled:
          type: "boolean"

security:
  - oauth_auth: []
```

## Context

I'll try to dig into why this is happening but I have not had a look at how speccy works as yet but I think it would be beneficial for other people who might want to use links as re-usable components.

## Your environment

Nothing really special to mention here a basic install of Node 8.10.0 (lts) on Linux Mint

```
Linux chris-H81M-S2V 4.13.0-36-generic #40~16.04.1-Ubuntu SMP Fri Feb 16 23:25:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by MikeRalphson** [comment]
Two things going on here:

* the `links` property of a `response Object` is defined as a `Map[string, Link Object | Reference Object]` so it can't itself be a `$ref`
* there is a bug/missing case in the validator where it is not dereferencing `$ref`s *within* `link Objects`

The fix to the first is to amend your second example to:

```yaml
  responses:
    204:
      description: "No content"
    404:
      description: "Resource not found"
    405:
      description: "Invalid input"
    410:
      description: "Resource has been deleted"
    administratorResponse:
      description: "The Administrator object response"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/Administrator"
      links:
        getUserById:
          $ref: "#/components/links/getUserById"
```

The second is for @philsturgeon to `git apply` this patch: https://github.com/Mermade/swagger2openapi/commit/d8c94d1122f88ff9f7a4ecb0a89da71a8bdda131.patch

Alternatively, add me as a collaborator, or I could PR it.

**Comment by catharsisjelly** [comment]
@MikeRalphson Thanks for your input on this, I have amended my example as you have requested, don't think I have the power to perform the other task

**Comment by philsturgeon** [comment]
You have the power, because a) you should, and b) I’m trapped in the mountains. 

—
Phil Sturgeon

**Comment by wol190** [comment]
I believe I'm experiencing the same issue with callbacks:

```
openapi: 3.0.0
servers:
  - url: 'http://127.0.0.1:8080'
info:
  title: 'Callback'
  version: '1'
  description: 'Callback test'
tags:
  - name: callbackEx
paths:
  /hooks:
    post:
      requestBody:
        $ref: '#/components/requestBodies/HooksPost'
      responses:
        '201':
          description: 'Successfully registered'
      callbacks:
        data_event:
          $ref: '#/components/callbacks/DataEvent'
components:
  callbacks:
    DataEvent:
      '{$request.body#/callbackUrl}':
        post:
          requestBody:
            $ref: '#/components/requestBodies/CallBackBody'
          responses:
            '200':
              description: 'Your callback should answer with 200'
  requestBodies:
    HooksPost:
        content:
          application/json:
            schema:
              type: object
              required:
               - callbackUrl
              properties:
                callbackUrl:
                  type: string
    CallBackBody:
      content:
        application/json:
          schema:
            required:
              - subscriptionId
            properties:
              subscriptionId:
                type: string
```
[Callback1RefExample.yaml.txt](https://github.com/wework/speccy/files/2996833/Callback1RefExample.yaml.txt)

```
> speccy lint .\Callback1RefExample.yaml
Specification contains lint errors: 1

#/components/callbacks/DataEvent  R: reference-no-other-properties  D: reference objects should only have a $ref property
expected Object {
  '{$request.body#/callbackUrl}': Object {
    post: Object {
      description: 'Callback event',
      operationId: 'Callbacks_POST',
  ... snip ...
} to have property $ref

More information: https://speccy.io/rules/1-rulesets#reference-no-other-properties
```

If I define the callbacks inline, there's no lint error:

```
      callbacks:
        data_event:
          '{$request.body#/callbackUrl}':
            post:
              requestBody:
                $ref: '#/components/requestBodies/CallBackBody'
              responses:
                '200':
                  description: 'Your callback should answer with 200'
```
[Callback1Example.yaml.txt](https://github.com/wework/speccy/files/2996853/Callback1Example.yaml.txt)

Confirmation of the issue (and ideally resolution) would be much appreciated. Excellent, tool - wish I'd come across it sooner!!!

**Comment by MikeRalphson** [comment]
@wol190 it would probably have been better to raise a new issue in order to track this, but no matter.

Fixed in oas-validator@3.2.3 commit https://github.com/Mermade/oas-kit/commit/62bbf0189f8f3f4ba17aa8482e445ce1244cd275

I can't comment on timescale for fixes propagating to speccy releases.

**Comment by wol190** [comment]
@MikeRalphson many thanks for the quick feedback - yup, I stumbled across this closed issue as I started typing a title for a new callback specific issue (so was torn with regards to starting a issue or not). Anyhow, good to hear that the way I've defined my callbacks is aligned with the spec (with either approach).
